### PR TITLE
ISPN-13991 Commands returning the CacheEntry

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-13991

Adding the commands for for `putAsync`, `replaceAsync`,
`putIfAbsentAsync`, and `removeAsync` on the `AdvancedCache`
returning a `CacheEntry`.

In this PR we are using the existing commands and an additional
boolean to know if it returns the value or the CacheEntry. Thought
this was the easiest approach for now.

Should we add specific commands that return the CacheEntry?
I dont like very much the approach right now because it mixes the
command return type, which could be complicated to maintain.
Not sure if always returning the CacheEntry and retrieving the 
command would be ok.

### CODE REVIEW

File: core/src/main/java/org/infinispan/cache/impl/EncoderCache.java [hunk 1]
```diff
@@ -347,6 +354,11 @@ public CompletableFuture<V> replaceAsync(K key, V value, Metadata metadata) {
       return cache.replaceAsync(keyToStorage(key), valueToStorage(value), metadata).thenApply(decodedValueForRead);
    }
 
+   @Override
+   public CompletableFuture<CacheEntry<K, V>> replaceAsyncEntry(K key, V value, Metadata metadata) {
+      return cache.replaceAsyncEntry(key, value, metadata);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/cache/impl/EncoderCache.java:359]: This is missing the unwrapping?

File: core/src/main/java/org/infinispan/cache/impl/EncoderCache.java [hunk 2]
```diff
@@ -945,4 +964,9 @@ private FunctionMapper wrapFunction(Function<?, ?> function) {
             null :
             new FunctionMapper(function, keyDataConversion, valueDataConversion);
    }
+
+   @Override
+   public CompletableFuture<CacheEntry<K, V>> removeAsyncEntry(Object key) {
+      return cache.removeAsyncEntry(key);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/cache/impl/EncoderCache.java:970]: This is missing the unwrapping?

File: core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java [hunk 3]
```diff
@@ -723,17 +728,22 @@ public V put(K key, V value, long lifespan, TimeUnit unit) {
...
+      ByRef<CacheEntry<K, V>> oldEntryRef = new ByRef<>(null);
       boolean hasListeners = this.hasListeners;
       getDataContainer().compute(key, (k, oldEntry, factory) -> {
          if (isNull(oldEntry)) {
             if (hasListeners) {
                CompletionStages.join(cacheNotifier.notifyCacheEntryCreated(key, value, metadata, true, ImmutableContext.INSTANCE, null));
             }
          } else {
-            oldRef.set(oldEntry.getValue(), oldEntry.getMetadata());
+            oldEntryRef.set(oldEntry.clone());
```
Comment (wburns) [inline core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:746]: Maybe just add a comment saying that we have to clone due to possibly updating the value reference below?

File: core/src/test/java/org/infinispan/api/AsyncAPITest.java [hunk 4]
```diff
@@ -91,6 +94,19 @@ public void testPutAsync() throws Exception {
       assertEquals("v2", c.get("k"));
    }
 
+   public void testPutAsyncEntry() throws Exception {
+      CompletableFuture<CacheEntry<String, String>> f = c.getAdvancedCache().putAsyncEntry("k", "v1", EmbeddedMetadata.EMPTY);
```
Comment (wburns) [inline core/src/test/java/org/infinispan/api/AsyncAPITest.java:98]: Would love to see this changed to use an actual metadata to assert as the client will be doing that on every write.

File: anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredDistributionInterceptor.java [hunk 5]
```diff
@@ -194,7 +194,7 @@ private VisitableCommand replaceWithPutRemoteMetadata(InvocationContext ctx, Dat
          }
 
          Metadata metadata = new RemoteMetadata(keyWriter, null);
-         PutKeyValueCommand copy = new PutKeyValueCommand(key, null, false, metadata,
+         PutKeyValueCommand copy = new PutKeyValueCommand(key, null, false, false, metadata,
```
[the remark this answers is not in the document]
Comment (jabolina) [reply anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredDistributionInterceptor.java]: Updated using the methods available in the factory, but it causes the command invocation id to change, is that a problem?

File: core/src/main/java/org/infinispan/cache/impl/CacheImpl.java [hunk 6]
```diff
@@ -1324,15 +1324,15 @@ private V putIfAbsent(K key, V value, Metadata metadata, long explicitFlags) {
...
    final V putIfAbsent(K key, V value, Metadata metadata, long explicitFlags, ContextBuilder contextBuilder) {
       assertKeyValueNotNull(key, value);
-      DataWriteCommand command = createPutIfAbsentCommand(key, value, metadata, explicitFlags);
+      DataWriteCommand command = createPutIfAbsentCommand(key, value, metadata, explicitFlags, false);
       return invocationHelper.invoke(contextBuilder, command, 1);
    }
 
-   private PutKeyValueCommand createPutIfAbsentCommand(K key, V value, Metadata metadata, long explicitFlags) {
+   private PutKeyValueCommand createPutIfAbsentCommand(K key, V value, Metadata metadata, long explicitFlags, boolean returnEntry) {
       long flags = addUnsafeFlags(explicitFlags);
```
[the remark this answers is not in the document]
Comment (jabolina) [reply core/src/main/java/org/infinispan/cache/impl/CacheImpl.java:1332]: IIUC the problem is that this method adds the `IGNORE_RETURN_VALUES`? Just to be sure before creating a JIRA.

File: anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredDistributionInterceptor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (wburns) [reply anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredDistributionInterceptor.java]: Ah, good point. Let me look at it closer, we may want to leave it how it was before then.

File: core/src/main/java/org/infinispan/cache/impl/CacheImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/cache/impl/CacheImpl.java:1332]: Correct, yes.

File: core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java [hunk 7]
```diff
@@ -808,26 +825,32 @@ public boolean replace(K key, V oldValue, V value, long lifespan, TimeUnit unit)
...
-      ValueAndMetadata<V> oldRef = new ValueAndMetadata<>();
+      ByRef<CacheEntry<K, V>> ref = new ByRef<>(null);
       boolean hasListeners = this.hasListeners;
       getDataContainer().compute(key, (k, oldEntry, factory) -> {
          if (!isNull(oldEntry)) {
             if (hasListeners) {
                CompletionStages.join(cacheNotifier.notifyCacheEntryModified(key, value, metadata, oldEntry.getValue(), oldEntry.getMetadata(), true, ImmutableContext.INSTANCE, null));
             }
-            oldRef.set(oldEntry.getValue(), oldEntry.getMetadata());
+            ref.set(oldEntry.clone());
```
Comment (wburns) [inline core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:842]: Make sure you put the same comment here as the clone above

File: core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java [hunk 8]
```diff
@@ -808,26 +825,32 @@ public boolean replace(K key, V oldValue, V value, long lifespan, TimeUnit unit)
...
       });
-      V oldValue = oldRef.getValue();
-      if (hasListeners && oldValue != null) {
+      CacheEntry<K, V> oldRef = ref.get();
+      if (hasListeners && oldRef != null && oldRef.getValue() != null) {
+         V oldValue = oldRef.getValue();
          CompletionStages.join(cacheNotifier.notifyCacheEntryModified(key, value, metadata, oldValue, oldRef.getMetadata(), false, ImmutableContext.INSTANCE, null));
       }
-      return oldValue;
+      return ref.get();
```
Comment (wburns) [inline core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:853]:
```suggestion
      return oldRef;
```

File: core/src/main/java/org/infinispan/commands/CommandsFactory.java [hunk 9]
```diff
@@ -169,7 +169,11 @@ public interface CommandsFactory {
     * @param flagsBitSet Command flags provided by cache
     * @return a PutKeyValueCommand
     */
-   PutKeyValueCommand buildPutKeyValueCommand(Object key, Object value, int segment, Metadata metadata, long flagsBitSet);
+   default PutKeyValueCommand buildPutKeyValueCommand(Object key, Object value, int segment, Metadata metadata, long flagsBitSet) {
+      return buildPutKeyValueCommand(key, value, false, segment, metadata, flagsBitSet);
+   }
+
+   PutKeyValueCommand buildPutKeyValueCommand(Object key, Object value, boolean returnEntry, int segment, Metadata metadata, long flagsBitSet);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/commands/CommandsFactory.java:176]: Missed moving the boolean to the end of the function like the replace one.

File: core/src/main/java/org/infinispan/commands/CommandsFactory.java [hunk 10]
```diff
@@ -179,7 +183,11 @@ public interface CommandsFactory {
     * @param flagsBitSet Command flags provided by cache
     * @return a RemoveCommand
     */
-   RemoveCommand buildRemoveCommand(Object key, Object value, int segment, long flagsBitSet);
+   default RemoveCommand buildRemoveCommand(Object key, Object value, int segment, long flagsBitSet) {
+      return buildRemoveCommand(key, value, false, segment, flagsBitSet);
+   }
+
+   RemoveCommand buildRemoveCommand(Object key, Object value, boolean returnEntry, int segment, long flagsBitSet);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/commands/CommandsFactory.java:190]: Missed moving the boolean to the end of the function like the replace one.

File: core/src/test/java/org/infinispan/api/AsyncAPITest.java [hunk 11]
```diff
@@ -91,6 +97,26 @@ public void testPutAsync() throws Exception {
       assertEquals("v2", c.get("k"));
    }
 
+   public void testPutAsyncEntry() throws Exception {
+      Metadata metadata = new EmbeddedMetadata.Builder().version(new NumericVersion(1)).build();
```
Comment (wburns) [inline core/src/test/java/org/infinispan/api/AsyncAPITest.java:101]: Metadata checks are good now, however can we have some diversity in the assertions with the versions and expiration? We are only ever inserting version 1 and no expiration and only asserting that. Having a different version and some different expiration times would be greatly appreciated to have some confidence it is working properly.

File: core/src/test/java/org/infinispan/api/AsyncAPITest.java [hunk 12]
```diff
@@ -91,6 +97,26 @@ public void testPutAsync() throws Exception {
...
+      assertEquals("v1", c.get("k"));
+
+      f = c.getAdvancedCache().putAsyncEntry("k", "v2", EmbeddedMetadata.EMPTY);
+      assertFutureResultOn(f, previousEntry -> {
+         assertEquals("k", previousEntry.getKey());
+         assertEquals("v1", previousEntry.getValue());
+         assertNotNull(previousEntry.getMetadata());
+
+         Metadata previousMetadata = previousEntry.getMetadata();
+         assertEquals(InequalVersionComparisonResult.EQUAL, previousMetadata.version().compareTo(new NumericVersion(1)));
```
Comment (wburns) [inline core/src/test/java/org/infinispan/api/AsyncAPITest.java:113]: Should be able to use just a simple `assertEquals(new NumericVersion(1), previousMetadata.version());` as they implement equality.

File: core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java [hunk 13]
```diff
@@ -778,15 +795,15 @@ protected V putIfAbsentInternal(K key, V value, Metadata metadata) {
             }
             return factory.create(k, value, metadata);
          } else {
-            oldValueRef.set(oldEntry.getValue());
+            previousEntryRef.set(oldEntry);
```
[the remark this answers is not in the document]
Comment (jabolina) [reply core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:798]: Feels like tests are missing.

[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:798]: It is a pretty small window/niche event. You would need an existing value and then a concurrent putIfAbsent and put while using the SimpleCache :D

[the remark this answers is not in the document]
Comment (jabolina) [reply core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java]: Added some simple tests for the SimpleCache.

File: core/src/test/java/org/infinispan/api/AsyncAPITest.java (hunk 11 above)
Comment (jabolina) [reply core/src/test/java/org/infinispan/api/AsyncAPITest.java]: The tests are using a metadata with lifespan and max idle and updating the values between operations.

File: core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java [hunk 14]
```diff
@@ -213,18 +215,27 @@ private Object performPut(MVCCEntry<Object, Object> e, InvocationContext ctx, Va
...
+            response = o;
+         }
+      }
+
       Metadatas.updateMetadata(e, metadata);
       if (e.isRemoved()) {
          e.setCreated(true);
          e.setExpired(false);
          e.setRemoved(false);
-         o = null;
```
[the remark this answers is not in the document]
Comment (jabolina) [reply core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java:222]: If the key is removed during the put I could return the wrong value, just updated to return null and keep the same behavior.

File: core/src/main/java/org/infinispan/cache/impl/CacheImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/cache/impl/CacheImpl.java:1333]: If that is the case, I would say make it a separate PR/JIRA

File: core/src/main/java/org/infinispan/commands/CommandsFactory.java [hunk 15]
```diff
@@ -226,17 +235,18 @@ public interface CommandsFactory {
...
+    * @param key         key to replace
+    * @param oldValue    existing value to check for if conditional, null if unconditional.
+    * @param newValue    value to replace with
+    * @param segment     the segment of the given key
+    * @param metadata    metadata of entry
     * @param flagsBitSet Command flags provided by cache
+    * @param returnEntry true if the {@link CacheEntry} is the command response, otherwise returns previous value.
     * @return a ReplaceCommand
     */
    ReplaceCommand buildReplaceCommand(Object key, Object oldValue, Object newValue, int segment, Metadata metadata,
```
Comment (wburns) [inline core/src/main/java/org/infinispan/commands/CommandsFactory.java:248]: Sorry if there is confusion here. But we should have the three commands have the similar overloads. Either we have no overload (1 method like this) or have the one overload (2 methods) like put and remove above.

**Comment by wburns** [comment]
LGTM, just waiting on CI

**Comment by wburns** [comment]
Integrated into main, thanks @jabolina !

**Pull request merged**